File: typescript/src/prompts/editor.ts
import { spawnSync } from "node:child_process";
import { closeSync, mkdtempSync, openSync, readFileSync, rmSync, writeSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { EditorError } from "../errors.js";
import { type BaseConfig, BasePrompt } from "./base.js";

export interface EditorConfig extends BaseConfig<string> {
  postfix?: string;
}

/**
 * Quote-aware shell-word split for $VISUAL/$EDITOR (R9). Honors single and
 * double quotes and backslash escapes; argv is exec'd WITHOUT a shell so there
 * is no injection surface. Mirrors the behavior of Python's shlex.split.
 */
export function splitCommand(input: string): string[] {
  const args: string[] = [];
  let current = "";
  let hasToken = false;
  let i = 0;
  const n = input.length;
  while (i < n) {
    const ch = input[i]!;
    if (ch === " " || ch === "\t" || ch === "\n" || ch === "\r" || ch === "\f" || ch === "\v") {
      if (hasToken) {
        args.push(current);
        current = "";
        hasToken = false;
      }
      i++;
      continue;
    }
    hasToken = true;
    if (ch === "'") {
      i++;
      while (i < n && input[i] !== "'") {
        current += input[i];
        i++;
      }
      i++; // skip closing quote
    } else if (ch === '"') {
      i++;
      while (i < n && input[i] !== '"') {
        if (input[i] === "\\" && i + 1 < n) {
          const next = input[i + 1]!;
          // In double quotes, backslash escapes only " \ $ ` (POSIX); otherwise literal.
          if (next === '"' || next === "\\" || next === "$" || next === "`") {
            current += next;
            i += 2;
            continue;
          }
        }
        current += input[i];
        i++;
      }
      i++; // skip closing quote
    } else if (ch === "\\" && i + 1 < n) {
      current += input[i + 1];
      i += 2;
    } else {
      current += ch;
      i++;
    }
  }
  if (hasToken) args.push(current);
  return args;
}

export class EditorPrompt extends BasePrompt<string> {
  private postfix: string;

  constructor(config: EditorConfig) {
    super(config);
    this.postfix = config.postfix ?? ".txt";
  }

  get promptType(): string {
    return "editor";
  }

  protected validateAnswer(value: unknown): string {
    if (value == null) return this.defaultValue ?? "";
    return String(value);
  }

  protected override toAgentDict(): Record<string, unknown> {
    return { ...super.toAgentDict(), postfix: this.postfix };
  }

  protected async executeTerminal(): Promise<string> {
    const editorCmd = process.env.VISUAL || process.env.EDITOR || "vi";
    const parts = splitCommand(editorCmd);
    const cmd = parts[0] ?? "vi";
    const cmdArgs = parts.slice(1);

    // Secure temp file: randomized directory (0700) + randomized file name
    // created with O_EXCL | O_CREAT (no symlink follow, no clobber) at mode
    // 0600, removed on EVERY exit path (R9).
    const dir = mkdtempSync(join(tmpdir(), "inquirer-"));
    const rand = Math.random().toString(36).slice(2) + Date.now().toString(36);
    const tmpPath = join(dir, `edit-${rand}${this.postfix}`);

    try {
      // "wx" = O_WRONLY | O_CREAT | O_EXCL — fails if the path already exists
      // and does not follow a symlink to clobber another file.
      const fd = openSync(tmpPath, "wx", 0o600);
      try {
        writeSync(fd, this.defaultValue ?? "");
      } finally {
        closeSync(fd);
      }

      const result = spawnSync(cmd, [...cmdArgs, tmpPath], { stdio: "inherit" });

      if (result.error) {
        throw new EditorError(`Editor not found: ${JSON.stringify(editorCmd)}. Set $VISUAL or $EDITOR.`);
      }
      if (result.status !== 0) {
        throw new EditorError(`Editor exited with code ${result.status}`);
      }

      return readFileSync(tmpPath, "utf8");
    } finally {
      // Remove the temp dir + file on success, editor-not-found, and non-zero exit.
      try {
        rmSync(dir, { recursive: true, force: true });
      } catch {
        // ignore cleanup errors
      }
    }
  }
}

File: rust/src/prompts/editor.rs
use crate::agent::agent_prompt_with_retry;
use crate::errors::{InquirerError, Result};
use crate::mode::is_agent_mode;
use crate::terminal::format_success;
use serde_json::{json, Value};
use std::io::{Read, Write};
use std::process::Command;

pub struct EditorConfig {
    pub message: String,
    pub default: Option<String>,
    pub postfix: String,
}

impl EditorConfig {
    pub fn new(message: impl Into<String>) -> Self {
        Self {
            message: message.into(),
            default: None,
            postfix: ".txt".to_string(),
        }
    }
}

pub fn editor(config: EditorConfig) -> Result<String> {
    if is_agent_mode() {
        editor_agent(&config)
    } else {
        editor_terminal(&config)
    }
}

fn editor_agent(config: &EditorConfig) -> Result<String> {
    let default = config.default.clone();
    let payload = json!({
        "type": "editor",
        "message": config.message,
        "default": config.default,
        "postfix": config.postfix,
    });

    agent_prompt_with_retry(&payload, move |answer| match answer {
        Value::Null => Ok(default.clone().unwrap_or_default()),
        Value::String(s) => Ok(s),
        other => Ok(other.to_string()),
    })
}

fn editor_terminal(config: &EditorConfig) -> Result<String> {
    let editor_cmd = std::env::var("VISUAL")
        .or_else(|_| std::env::var("EDITOR"))
        .unwrap_or_else(|_| "vi".to_string());

    // Quote-aware shell-word split, then exec WITHOUT a shell (no injection).
    let parts = shell_words::split(&editor_cmd)
        .map_err(|e| InquirerError::Editor(format!("Invalid $EDITOR/$VISUAL: {e}")))?;
    let (cmd, args) = parts
        .split_first()
        .ok_or_else(|| InquirerError::Editor("Empty EDITOR command".into()))?;

    // Secure temp file: randomized name, mode 0600, created with O_EXCL (no
    // symlink follow, no clobber), removed on EVERY exit path via Drop.
    let mut named = build_temp_file(&config.postfix)?;

    if let Some(d) = &config.default {
        named
            .as_file_mut()
            .write_all(d.as_bytes())
            .map_err(InquirerError::Io)?;
        named.as_file_mut().flush().map_err(InquirerError::Io)?;
    }

    let status = Command::new(cmd)
        .args(args)
        .arg(named.path())
        .status()
        .map_err(|_| {
            InquirerError::Editor(format!(
                "Editor not found: {editor_cmd:?}. Set $VISUAL or $EDITOR."
            ))
        })?;

    if !status.success() {
        // `named` is dropped here, removing the temp file.
        return Err(InquirerError::Editor(format!(
            "Editor exited with code {}",
            status.code().unwrap_or(-1)
        )));
    }

    // Re-read from the start: the editor may have rewritten the file.
    let mut content = String::new();
    {
        let mut f = std::fs::File::open(named.path()).map_err(InquirerError::Io)?;
        f.read_to_string(&mut content).map_err(InquirerError::Io)?;
    }
    // `named` is dropped at end of scope, removing the temp file on success too.

    eprintln!("{}", format_success(&config.message, "(editor)"));
    Ok(content)
}

/// Create a secure temporary file with a randomized name, mode 0600, created
/// with O_EXCL (no clobber, no symlink follow), using the configured postfix
/// as the file suffix.
fn build_temp_file(postfix: &str) -> Result<tempfile::NamedTempFile> {
    let mut builder = tempfile::Builder::new();
    builder.prefix("inquirer-edit-").suffix(postfix);
    let named = builder
        .tempfile()
        .map_err(|e| InquirerError::Editor(format!("Failed to create temp file: {e}")))?;
    // tempfile creates with O_EXCL and 0600 on unix; set the mode explicitly
    // to be robust against umask differences.
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        named
            .as_file()
            .set_permissions(std::fs::Permissions::from_mode(0o600))
            .map_err(InquirerError::Io)?;
    }
    Ok(named)
}
